### DISCUSSION TITLE
Strange behavior during thread cutting

### DISCUSSION THREAD

**Discussion by duebelpapst** [kachurovskiy/nanoels#151] [General]
Hi

Today I tried to cut a M12 thread and I had a very strange behavior.

Limit values Z10 / Z0; X-6 / X-4.7 
Passes 7
Pitch 1.75
Speed 300 rpm

During thread cutting, Z moves slowly at first, then speeds up and then slows down again so that the thread is completely miscut.
I thought the speed would drop under load or something, so I tested the whole thing several times without material, same result.
Z also moves a few tenths further towards the chuck with each pass, but the display is always Z10, on the far left.
Before tapping I marked the chuck at an angle of 0°, after tapping the marking did not match the displayed angle.
I connected the encoder screen to minus.
What confuses me is that the stepper motor moves farther than indicated.

FYI, I started building my own 2 axis ELS a long time ago, and have tested the actual hardware extensively, including threading, and everything has always worked.
I built the whole thing with an Arduino Mega plus Nano for the displays, on a breadboard.
I recently came across the NanoELS H4 and thought that for the money I could save myself the trouble of building my own board.

By the way, the stepper motor is a JMC IHSS57-36-20, the encoder is a [WISAMIC 600p/r](https://www.amazon.de/gp/product/B015GYY7XU/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1
)

This evening I will supply the encoder with external 5V, but I don't think this will solve the problem.

Does anyone has an idea?

Kind regards,
Stefan

**Comment by Festivejelly** [comment]
I was having issues on my Z axis not moving the correct amount. Even though I have an open loop driver and motor I set the config value

NEEDS_REST_Z = false

As soon as I did that the accuracy issues went away. You might be facing a similar issue. Worth trying it.

**Reply by duebelpapst** [reply]
Hello,
"false" is standard, I haven't changed it.
But maybe the stepper is getting some extra pulses due to "crosstalk" on the step-wire.
I will check that.

**Comment by kachurovskiy** [comment]
Hi @duebelpapst, sorry that you're facing issues. Let's go through a checklist and hopefully we can pinpoint the problem.

Please post your entire h4.ino on https://gist.github.com/ and share the link with us here.

1. Turn on angle display, press Stop to zero the angle, note the chuck position and turn it exactly 1 turn. Angle on display should change by 360 back to 0. If it does, try turning on the lathe motor, making a few dozen chuck turns, stopping the motor, rotating the chuck forward until it hits the start position and checking that angle shown on screen is also 0. If any of these tests fail, we have a problem with encoder pulses - either it needs move current or more voltage or there's noise on the signal lines.
2. Setup a dial indicator to see the Z axis movements, set `Step` to e.g. `5mm` and use manual move commands Left and Right to check if you're coming back to the same place. Try doing this with main lathe motor enabled and disabled to check if main motor / VFD noise is interfering with the stepper signal.

If none of this helps, please share a short video of the problem, that might give us some hints. Thanks!

**Comment by duebelpapst** [comment]
Hi Maxim,

I've already checked a few things.
The problem seems to be the timing of the encoder and display.
I had a similar problem with my one ELS, that's the reason I'm using a Mega with a Nano.
But I thought that an ESP32 can handle this problem.
In the code I can't find a way to change the interval for updating the display.

Today I changed a few things before I do the tests:
1. changed both steps per rev to 400, Z was 1600.
2. changed the encoder w/ 600 pulses to one w/ 360 pulses
3. ext. 5V for the encoder
4. both steppers are disconnected
 
When I turn the chuck by hand it shows for every revolution 360°.
Then I tried 2000 rpm for a aprox. 10 seconds, totaly off track.
1000 rpm a bit better.
300 rpm almost good.

What I also did is, I 0 the angle, turned off the angle display, turned the chuck w/ the motor, stopped it, turned the angle display on, but it doesn't help.

Than I connected the stepper, and cut a M10 thread, 300 rpm, 7 passes, L=10, angle = 0.
After the cut the angle = 184.
I still can notice a speed change of z during the cut.

Her is the link to my code.
[duebelpapst](https://gist.github.com/duebelpapst)/[duebelpapst_nanoels_ h4.ino](https://gist.github.com/duebelpapst/31c14f5979798f7807687206e02cc837)

Kind regards,
Stefan

**Reply by kachurovskiy** [reply]
I'm afraid updating the display has nothing to do with it since it's done on another core and encoder pulses are handled as interrupts anyway.

1. Can you please post some pictures of your machine, especially the encoder?
2. Using a belt on encoder instead of gears could cause such problems, I assume you use gears?
3. If you have some small ceramic capacitors, can you try installing them between the encoder signal lines and GND or 5V? This might filter out the noise that could register as extra pulses.

**Reply by duebelpapst** [reply]
1. I will post the pics tomorrow. I tried 3 encoder, 200; 360 and 600
2. I use a gear
3. yes i have 1nF, 10nF

**Reply by duebelpapst** [reply]
Now that you metioned the capasitors I remember that I use pull up resistors at the signal lines at my ELS. I try this tomorrow too.

**Reply by kachurovskiy** [reply]
I have pull-up resistors on the PCB for the encoder signal lines, no need to add those externally.

**Reply by duebelpapst** [reply]
Ok

**Reply by duebelpapst** [reply]
Hi Maxim,

it's going worse.
I connected the two capacitors between the signal lines of the 360 encoder and GND.
in this case the encoder / display shows only approx. 240° per revolution.
Then I tried it w/ 5V, now it shows only approx. 52° per rev.
The same happend w/ the 600 encoder, settings in the prog are changed to 600.

I also tried it w/ ext. power and power from the ELS.
It doesn't matter, same result.

Even when I turned the encoder clockwise and back, it goes not to the same angle.

By the way, when I use ext. power, I only connected the GND to the ELS, the 5V I only connected to the encoder, is this OK, or should I connect also the 5V to the ELS?

The main motor was always off.
Some pics are attached.
1. 360 encoder mounted to the spindel
2. 600 encoder
3. z-axis stepper
4. x-axis stepper
5. connectors
6. capacitors I use

Regards,
Stefan
![20231215_140130](https://github.com/kachurovskiy/nanoels/assets/45849605/ced6a26b-5bb4-4b6a-bd36-9da32ebcc884)
![20231215_140140](https://github.com/kachurovskiy/nanoels/assets/45849605/dabb051c-d38c-49da-b24f-11392104e42e)
![20231215_140153](https://github.com/kachurovskiy/nanoels/assets/45849605/e7482502-a972-4c62-bca9-193cd66ca289)
![20231215_140202](https://github.com/kachurovskiy/nanoels/assets/45849605/b7000715-c96a-4e4e-8b56-5a8db920382d)
![20231215_140216](https://github.com/kachurovskiy/nanoels/assets/45849605/27634142-3c23-446a-b6c2-1d757e20cae5)
![20231215_140335](https://github.com/kachurovskiy/nanoels/assets/45849605/9f386d43-5fd0-4aa8-ae4b-ad40a274b491)

**Reply by duebelpapst** [reply]
Hi Maxim,

I've tried a few other things and it looks like the VFD or main motor is part of the problem.
When I turn the chuck by hand or with a cordless screwdriver while tapping (without material), everything seems to work fine except for the angle.
When I rotate the chuck with the cordless screwdriver, the positioning of the Z-axis is fine, it stops at both ends of the travel at the intended point.
But the angle is still irregular.

With the main motor, the z-axis position moves 0.1 to 0.4mm to the chuck on each pass, it depends on the spindle speed, the X-axis moves 0.005 to 0.01 mm to -X. 

But I don't understand why the VFD works on my own ELS and not on the nanoels.

I have changed the mains filter [this is the filter](https://www.sorotec.de/shop/advanced_search_result.php?categories_id=0&keywords=ZFU.NF1024SE.LL&inc_subcat=1) of the lathe, and also the USB cable, nothing has changed.

Can I supply the nanoels with power via the GND and 5V of the encoder connector?

At the moment I give up, and switched back to my own ELS, everything else is the same, and it works.

Video 1 nanoels H4, can you see the acceleration?
Video 2 my own, the noise comes from the X-Stepper, I forgot to change the steps on the driver.

Do you have an idea whats the problem?


https://github.com/kachurovskiy/nanoels/assets/45849605/75f27e50-53c1-4a29-a6e9-061542e0d574

https://github.com/kachurovskiy/nanoels/assets/45849605/d5bdeb5c-3525-4208-b04f-41a05a262863

Regards,
Stefan

**Reply by kachurovskiy** [reply]
The initial acceleration on H4 can sound weird in gearbox/threading modes because it tries to catch up with the spindle for the first second or so. After that is slows down a bit and goes at constant speed. The idea is to account for this by leaving some room between the part and the tool for acceleration and/or running the motor a little slower on large pitches.

It seems like VFD causes excessive pulses on encoder and stepper signal lines. Are the X and Z signal wires from H4 to stepper driver shielded and connected to 5V?

When you start addressing the noise issues from the VFD, a good way to check the progress would be disengaging the encoder from the spindle e.g. by moving it a bit out of the way and turning the VFD on. If H4 angle display shows any angle changes, it's not fixed yet.

**Reply by duebelpapst** [reply]
Dear Maxim,

shield of the steppers are connected to GND.
I have already plugged the USB power supply into a different phase.
What just confuses me is that the VFD, with the line filter, doesn't mess anything up with my hardware, which is just patched together on a breadboard.
I will test what you suggested.  I'll also turn on the motor and turn the encoder, away from the motor, with the cordless drill. We will see what happens.
Regards,
Stefan

**Reply by kachurovskiy** [reply]
The EMI noise is unlikely to come from the USB power - H4 has filters there - instead it's coming through the air. Your breadboard Nano/Mega setup is likely more resistant to it because you're driving everything directly via I/O pins which allow for more current on the signal lines making it harder for the EMI noise to mess with the signal.

**Reply by duebelpapst** [reply]
That makes sense. But at the moment I have no idea how to solve this. Unfortunately my housing for the VFD is made of plastic. The funny thing is, my breadboard is very close to the VFD in this housing.

**Reply by duebelpapst** [reply]
I have tried the following:
Only the Z-stepper and the encoder connected, both shields to GND.
Encoder not on the spindle and turned by hand, approx. 1m away from the VFD.
When the VFD / motor is off, everything is OK.
When the VFD / motor is on, I have turned the encoder 20 turns by hand, and the angle is exactly right, so the VFD probably has no influence on the encoder.
But when I have the VFD / motor on and thread cutting, every pass the pitch is still shifted by a few tenths.
Next I will disconnect the motor from the VFD, see if this makes a diffrence.

**Reply by duebelpapst** [reply]
Hello Maxim,

when I disconnected the motor cable, everything was OK even with the VDF switched on.
Next I connected other motor, but placed it next to the lathe earth / ground isolated from the lathe, also OK.
Then I connected the housing of the motor to the housing of the lathe, and immediately I had an offset of 0.4mm per gear again.
I then connected the motor cable in the control box directly to the VFD and not via the terminal blocks, only the earth / ground was still on the terminal blocks, so it was still bad.
Only when I connected the earth / ground cable as short as possible directly to the VFD was it OK, sometimes I still have 0.01mm offset, but I think I can manage that.
By the way, it still gets really bad when I connect the shield of the Z-stepper to GND or +5V, the stepper has no ground connection to the housing of the lathe.
Now I will buy a metal switch cabinet and a better shielded motor cable, and hope that I can then get everything under control.
Maybe I have too many ground loops.
I should know it better, 42 years ago I learned to be an electrician, but since 25 years I am out of this buisness, now it is only a hobby.

Thanks for your help.
Regards,
Stefan

**Reply by kachurovskiy** [reply]
Stefan, happy that you've figured it out, when fully resolved please consider posting a summary with recommendations for other VFD users who will face similar problems in the future! Thank you.

**Reply by duebelpapst** [reply]
That was the plan

**Comment by duebelpapst** [comment]
Hi,

After several hours (approx. 60h) of trial and error, I finally seem to have found a solution.
Something seems to be interfering with the JMC closed loop stepper, unfortunately I couldn't find or fix the cause.
I tried the following:
replaced all the cables with shielded ones
all parts of the lathe have been earthed
bought a new metal housing
put additional ferrite rings over the individual wires of the motor supply cable.
tested 3 different mains filters at the input of the VFD
tested a different VFD
tested a different JMC
none of this produced a reproducible solution / improvement.

I still had additional steps on the Z stepper in the direction of the chuck when the spindle motor was running, but not in the other direction.
On X everything was OK, I had a stepper from Stepperonline with a cheap TB6600 driver.
My suspicion was that some interference signal is superimposed on the pulse/step signal line, and the voltage sometimes drops below the level that the JMC needs as high, and the JMC interprets this as a pulse. Why only in one direction? Perhaps there is also a connection with the DIR signal, but I don't know.
This is just an guess, as I don't have an oscilloscope I can't check it.
Now I was hoping to stabilize the signal from the Nanoels H4 with an optocoupler or ULN2803.
It didn't work with an optocoupler, it was much too slow.
But with an ULN2803 it worked immediately, I took GND from the encoder plug, 5V are already at the Z-stepper plug.
After extensive testing, everything now works as it should.
I have replaced the metal housing with the original plastic one and it still works.

Thanks to Maxim for the discussion and the suggestions to solve the problem.

Regards,
Stefan

**Reply by kachurovskiy** [reply]
Stefan, sorry that you had to go through all this trouble. Can you please share some pictures of your final wiring for us to understand how exactly you fixed the problem? Many thanks 🙏
